python/cuml/test/test_fil.py

```diff
@@ -492,29 +492,39 @@ def test_output_args(small_classifier_and_preds):
 
 
 def to_categorical(features, n_categorical):
+    # returns data in two formats: pandas (for LightGBM) and numpy (for FIL)
     # the main bottleneck (>80%) of to_categorical() is the pandas operations
     n_features = features.shape[1]
     df_cols = {}
     # all categorical columns
     cat_cols = features[:, :n_categorical]
-    cat_cols = cat_cols - cat_cols.min(axis=1, keepdims=True)  # range [0, ?]
-    cat_cols /= cat_cols.max(axis=1, keepdims=True)  # range [0, 1]
+    cat_cols = cat_cols - cat_cols.min(axis=0, keepdims=True)  # range [0, ?]
+    cat_cols /= cat_cols.max(axis=0, keepdims=True)  # range [0, 1]
     rough_n_categories = 100
     # round into rough_n_categories bins
     cat_cols = (cat_cols * rough_n_categories).astype(int)
+    new_features = features.copy()
+    new_features[:, :n_categorical] = cat_cols
+
+    # shuffle the columns around
+    seed(42)
+    new_idx = sample(range(n_features), k=n_features)
+    new_matrix = new_features[:, new_idx]
     for icol in range(n_categorical):
         col = cat_cols[:, icol]
+        # randomly inject invalid categories
+        n_rows = features.shape[0]
+        invalid_idx = sample(range(n_rows), k=n_rows // 10 + 1)
```

Comment:
It's better to pass the number or fraction (per column) of rows that get an invalid category value as a function parameter. It could be even better to pass the fraction for the whole sub-array of categorical columns.